From: Gavin Lambert <gavinl@compacsort.com>
Date: Wed, 1 Jun 2016 10:01:30 +0200

Update newly added text for redundancy device names.

diff --git a/master/master.c b/master/master.c
--- a/master/master.c
+++ b/master/master.c
@@ -3019,7 +3019,8 @@ int ecrt_master_get_slave(ec_master_t *m
     }
 
     if (slave->sii_image == NULL) {
-        EC_MASTER_WARN(master, "Cannot access SII data from slave position %u", slave->ring_position);
+        EC_MASTER_WARN(master, "Cannot access SII data from slave position %s-%u",
+                ec_device_names[slave->device_index!=0], slave->ring_position);
         ret = -ENOENT;
         goto out_get_slave;
     }
