From: Gavin Lambert <gavinl@compacsort.com>
Date: Mon, 11 Jul 2016 12:29:32 +1200

Adds slave position to ecrt_slave_config_state.

diff --git a/include/ecrt.h b/include/ecrt.h
--- a/include/ecrt.h
+++ b/include/ecrt.h
@@ -322,6 +322,7 @@ typedef struct  {
                                  bit! */
     unsigned int error_flag : 1; /**< The slave has an unrecoverable error. */
     unsigned int ready : 1; /**< The slave is ready for external requests. */
+    uint16_t position; /**< Offset of the slave in the ring. */
 } ec_slave_config_state_t;
 
 /*****************************************************************************/
diff --git a/master/slave_config.c b/master/slave_config.c
--- a/master/slave_config.c
+++ b/master/slave_config.c
@@ -1267,11 +1267,13 @@ void ecrt_slave_config_state(const ec_sl
         state->al_state = sc->slave->current_state;
         state->error_flag = sc->slave->error_flag ? 1 : 0;
         state->ready = ec_fsm_slave_is_ready(&sc->slave->fsm) ? 1 : 0;
+        state->position = sc->slave->ring_position;
     } else {
         state->operational = 0;
         state->al_state = EC_SLAVE_STATE_UNKNOWN;
         state->error_flag = 0;
         state->ready = 0;
+        state->position = (uint16_t) -1;
     }
 }
 
